aws/src/main/java/org/apache/iceberg/aws/sns/SNSListener.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.aws.sns;
+
+import org.apache.iceberg.events.Listener;
+import org.apache.iceberg.util.EventParser;
+import org.apache.iceberg.util.Tasks;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.services.sns.SnsClient;
+import software.amazon.awssdk.services.sns.model.NotFoundException;
+import software.amazon.awssdk.services.sns.model.PublishRequest;
+import software.amazon.awssdk.services.sns.model.SnsException;
+
+public class SNSListener<T> implements Listener<T> {
+  private static final Logger LOG = LoggerFactory.getLogger(SNSListener.class);
+
+  private final String topicArn;
+  private final SnsClient sns;
+  private final int retry;
+  private final long retryIntervalMs;
+
+  public SNSListener(String topicArn, SnsClient sns, int retry, long retryIntervalMs) {
+    this.sns = sns;
+    this.topicArn = topicArn;
+    this.retry = retry;
+    this.retryIntervalMs = retryIntervalMs;
+  }
+
+  @Override
+  public void notify(Object event) {
+    try {
+      String msg = EventParser.toJson(event);
+      PublishRequest request = PublishRequest.builder()
+              .message(msg)
+              .topicArn(topicArn)
+              .build();
+      Tasks.foreach(request)
+          .exponentialBackoff(retryIntervalMs, retryIntervalMs, retryIntervalMs, 1 /* scale factor */)
+          .retry(retry)
+          .onlyRetryOn(NotFoundException.class)
+          .run(sns::publish);
```

Comment:
this call is synchronous, we need to use the asynchronous client to make non-blocking call. Or should the listener notification process itself be non-blocking? Maybe that is a better approach to go to avoid taking to long to notify all listeners. Overall looks like we need a base Listener class to do:
1. asynchronous notificaiton
2. error handling to avoid failing a scan/create snapshot due to notification failure
3. retry in case of certain failures

Reply:
I'd be up for adding error handling to `Listeners.notifyAll`. That should probably never fail.

For async notification, I think that's a matter of starting a thread and adding a concurrent queue, so that should be done in the listeners themselves. Maybe we could create a base class for it though.

Reply:
+1 for a base class